pvlib/irradiance.py

```diff
@@ -833,7 +838,33 @@ def haydavies(surface_tilt, surface_azimuth, dhi, dni, dni_extra,
     sky_diffuse = dhi * (AI * Rb + term1 * term2)
     sky_diffuse = np.maximum(sky_diffuse, 0)
 
-    return sky_diffuse
+    # we've preserved the input type until now, so don't ruin it!
+    # if isinstance(sky_diffuse, pd.Series):
+    #     sky_diffuse[np.isnan(surface_tilt)] = 0
+    # else:
+    #     sky_diffuse = np.where(np.isnan(surface_tilt), 0, sky_diffuse)
+
+    if return_components:
+        diffuse_components = OrderedDict()
+        diffuse_components['sky_diffuse'] = sky_diffuse
+
+        # Calculate the individual components
+        diffuse_components['isotropic'] = np.maximum(dhi * term1 * term2, 0)
+        diffuse_components['circumsolar'] = np.maximum(dhi * (AI * Rb), 0)
+        diffuse_components['horizon'] = np.where(
+            np.isnan(diffuse_components['isotropic']), np.nan, 0.)
+
+        # Set values of components to 0 when sky_diffuse is 0
+        mask = sky_diffuse == 0
```

Comment:
I suspect this mask will never actually change any behavior, since the only way that `sky_diffuse` is zero is if all the components are already zero anyway.  Is that true?  If so, let's ditch `mask`.